R/valley.R

```diff
@@ -0,0 +1,299 @@
+#' @param crs CRS to be projected to
+#' @param ... Optional arguments for raster or vector reproject functions
+#'
+#' @return Object reprojected to specified CRS
+#' @export
+reproject <- function(x, crs, ...) {
+  if (inherits(x, "SpatRaster")) {
+    return(terra::project(x, crs, ...))
+  } else if (inherits(x, c("bbox", "sfc", "sf"))) {
+    return(sf::st_transform(x, crs, ...))
+  } else {
+    stop(sprintf("Cannot reproject object type: %s", class(x)))
+  }
+}
+
+#' spatially smooth dem by (window) filtering
+#'
+#' @param dem raster data of dem
+#' @param method smoothing function to be used, e.g. "median".
+#' As accepted by focal
+#' @param window size of smoothing kernel
+#'
+#' @return filtered dem
+#' @export
+smooth_dem <- function(dem, method = "median", window = 5) {
+  dem_smoothed <- terra::focal(dem, w = window, fun = method)
+  names(dem_smoothed) <- "dem_smoothed"
+  return(dem_smoothed)
+}
+
+#' Derive slope as percentage from dem
+#' This makes use of the terrain function of the terra package
+#'
+#' @param dem raster data of dem
+#'
+#' @return raster of derived slope over dem extent
+#' @export
+get_slope <- function(dem) {
+  slope_radians <- terra::terrain(dem, v = "slope", unit = "radians")
+  slope <- tan(slope_radians)
+  return(slope)
+}
+
+#' Mask slope raster, setting the slope to zero for the pixels overlapping
+#' the river area.
+#'
+#' @param slope raster data of slope
+#' @param vector/polygon data of river
+#' @param lthresh lower numerival threshold to consider slope non-zero
+#'
+#' @return updated slope raster
+#' @export
+mask_slope <- function(slope, river, lthresh = 1.e-3, target = 0) {
+  slope_masked <- terra::mask(
+                              slope,
+                              terra::ifel(slope <= lthresh, NA, 1),
+                              updatevalue = lthresh)
+
+  slope_masked <- terra::mask(
+                              slope_masked,
+                              river,
+                              inverse = TRUE,
+                              updatevalue = target,
+                              touches = TRUE)
+}
+
+#' Derive cost distance function from masked slope
+#'
+#' @param slope_masked raster of masked slope data
+#' @param target value for codt distance calculation
+#'
+#' @return raster of cost distance
+#' @export
+get_cost_distance <- function(slope, river, target = 0) {
+  slope_masked <- CRiSp::mask_slope(slope, river, target = target)
+  cd <- terra::costDist(slope_masked, target = target)
+  names(cd) <- "cost_distance"
+  return(cd)
+}
+
+#' Mask out river regions incl. a buffer in cost distance raster data
+#'
+#' @param cd cost distance raster
+#' @param river vector/polygon
+#' @param BUFFER size of buffer around river polygon to additionally mask
+#'
+#' @return cd raster with river+BUFFER pixels masked
+#' @export
+mask_cost_distance <- function(cd, river, buffer = 2000) {
+  river_buffer <- sf::st_buffer(river, buffer)
```

Comment:
```suggestion
  river_buffer <- sf::st_buffer(river, buffer)  |> terra::vect()
```
This is needed so that `cd` and `river_buffer` are compatible in `terra::mask()`